AntiResolvers

richard-rogers opened the pull request:
## Description

Creates `AntiResolver <: ResolverSpec` that removes the specified metrics from the set produced by `DeclarativeResolver::resolve()`. Allows `@register_dataset_udf` decorator to suppress specified metrics on the UDF output columns.

```
@register_dataset_udf(["col1"], "annihilate_me", anti_metrics=[CardinalityMetric, DistributionMetric])
def plus1(x):
    return x["col1"] + 1

def test_anti_resolver() -> None:
    schema = udf_schema()
    data = pd.DataFrame({"col1": [42, 12, 7], "col2": ["a", "b", "c"]})
    results = why.log(pandas=data, schema=schema).view()
```

The `annihilate_me` column will have the metrics specified by `STANDARD_RESOLVER` except `CardinalityMetric` and `DistributionMetric`.


- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

richard-rogers committed:
AntiResolvers

richard-rogers committed:
dont reset udfs

richard-rogers committed:
Update python/tests/experimental/core/test_udf_schema.py

richard-rogers committed:
Update python/tests/experimental/core/metrics/test_udf_metric.py

richard-rogers committed:
pre-commit

FelipeAdachi commented on python/tests/experimental/core/test_udf_schema.py:
```diff
@@ -14,6 +14,8 @@
     udf_schema,
 )
 
+# _reset_udfs()
```
same as above

FelipeAdachi commented on python/whylogs/core/resolvers.py:
```diff
@@ -109,7 +109,7 @@ class ResolverSpec:
 
     column_name: Optional[str] = None  # TODO: maybe make this a regex
-    column_type: Optional[Any] = None
+    column_type: Optional[type] = None
```
isn't `column_type` expected to be a DataType?

richard-rogers replied:
```suggestion
    column_type: Optional[DataType] = None
```

richard-rogers committed:
Merge branch 'mainline' into dev/richard/anti

richard-rogers committed:
Update python/whylogs/core/resolvers.py

richard-rogers committed:
Merge branch 'mainline' into dev/richard/anti

richard-rogers committed:
Merge branch 'mainline' into dev/richard/anti

richard-rogers committed:
Replace AntiResolver with field in ResolverSpec (#1280)

## Description

Adds a boolean field to `ResolverSpec` to exclude a metric rather than
add it.



- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md)
and the [Code of Conduct](CODE_OF_CONDUCT.md).

jamie256 reviewed (dismissed):
LGTM!

richard-rogers committed:
Update resolvers.py

doc string

richard-rogers committed:
Update udf_metric.py

doc string

richard-rogers commented on python/whylogs/experimental/core/metrics/udf_metric.py:
```diff
@@ -28,36 +28,24 @@
     match and the list of MetricSpecs to instantiate for matching UDFs.
     Each MetricSpec specifies the Metric class and MetricConfig to
     instantiate. Omit the MetricSpec::config to use the default MetricConfig.
+    ResolverSpec::exclude will exclude the listed metrics from the matched UDFs.
```
```suggestion
    Setting ResolverSpec::exclude to True will exclude the listed metrics from the matched UDFs.
```

richard-rogers committed:
Update python/whylogs/experimental/core/metrics/udf_metric.py

richard-rogers committed:
Update udf_schema.py

richard-rogers committed:
Update python/tests/experimental/core/metrics/test_udf_metric.py

richard-rogers committed:
Update python/tests/experimental/core/test_udf_schema.py

richard-rogers committed:
Update python/tests/experimental/core/test_udf_schema.py

richard-rogers committed:
Merge branch 'mainline' into dev/richard/anti

richard-rogers committed:
merge cleanup

richard-rogers committed:
Merge branch 'mainline' into dev/richard/anti

jamie256 reviewed (approved):
LGTM!
